mailpoet/lib/Automation/Engine/Control/SubjectTransformerHandler.php

```diff
@@ -21,181 +19,85 @@ public function __construct(
+    $triggerData = array_values(array_filter(
       $automation->getSteps(),
       function (StepData $step): bool {
         return $step->getType() === StepData::TYPE_TRIGGER;
       }
-    );
+    ));
 
     $triggers = array_filter(array_map(
-      function(StepData $step): ?Step {
-        return $this->registry->getStep($step->getKey());
+      function (StepData $step): ?Trigger {
+        return $this->registry->getTrigger($step->getKey());
       },
-      $configuredTriggers
+      $triggerData
     ));
-
-    $triggerKeys = [];
+    $all = [];
     foreach ($triggers as $trigger) {
-      $triggerKeys[$trigger->getKey()] = $trigger->getSubjectKeys();
-    }
-    if (!$triggerKeys) {
-      return [];
+      $all[] = $this->getSubjectKeysForTrigger($trigger);
     }
-
-    $triggerKeys = count($triggerKeys) === 1 ? current($triggerKeys) : array_intersect(...array_values($triggerKeys));
-    $possibleKeys = [];
-    foreach ($triggerKeys as $key) {
-      $possibleKeys = array_merge($possibleKeys, $this->getPossibleTransformations($key));
-    }
-    return array_unique(array_values(array_merge($triggerKeys, $possibleKeys)));
+    $all = count($all) > 1 ? array_intersect(...$all) : $all[0] ?? [];
+    return array_values(array_unique($all));
   }
 
-  private function getPossibleTransformations(string $key, string $stopKey = null): array {
-    if (!$stopKey) {
-      $stopKey = $key;
+  public function getSubjectKeysForTrigger(Trigger $trigger): array {
+    $all = [];
+    foreach ($trigger->getSubjectKeys() as $key) {
+      $all = array_merge($all, $this->getSubjectKeysForSingleKey($key));
     }
-    $allTransformer = $this->registry->getSubjectTransformer();
-    $possibleTransformer = array_filter(
-      $allTransformer,
-      function (SubjectTransformer $transformer) use ($key): bool {
-        return $transformer->accepts() === $key;
-      }
-    );
+    return $all;
+  }
 
-    $possibleKeys = [];
-    foreach ($possibleTransformer as $transformer) {
-      $possibleKey = $transformer->returns();
-      if ($possibleKey === $stopKey) {
-        continue;
-      }
-      $possibleKeys[$possibleKey] = $possibleKey;
-      $possibleKeys = array_merge(
-        $possibleKeys,
-        $this->getPossibleTransformations($possibleKey, $stopKey)
-      );
+  public function getAllSubjects(Subject ...$subjects): array {
+    $transformerMap = $this->getTransformerMap();
+    $all = [];
+    foreach ($subjects as $subject) {
+      $all[$subject->getKey()] = $subject;
     }
-    return $possibleKeys;
-  }
 
-  /**
-   * @param Trigger $trigger
-   * @param Subject ...$subjects
-   * @return Subject[]
-   */
-  public function provideAllSubjects(Trigger $trigger, Subject ...$subjects): array {
-    $allSubjectsKeys = $this->subjectKeysForTrigger($trigger);
-    $allSubjectKeyTargets = array_filter(
-      array_diff($allSubjectsKeys, array_map(
+    $queue = array_map(
```

Comment:
I wonder why you didn't keep $queue = array_keys($all), which seems to be much shorter.

Reply:
Right! 1db8d564a